File: apps/backend/app/schemas/resume.py
from pydantic import BaseModel, Field
from typing import List, Optional, Dict, Any
from uuid import UUID

class PersonalInformationSchema(BaseModel):
    name: str = ""
    email: str = ""
    phone: str = ""
    linkedin: str = ""
    github: str = ""
    portfolio: str = ""
    location: Optional[str] = ""
    headline: Optional[str] = ""

class EducationEntrySchema(BaseModel):
    id: Optional[str] = None
    degree: str = ""
    college: str = ""
    cgpa: str = ""
    graduationYear: str = ""

class CategorizedSkillsSchema(BaseModel):
    programmingLanguages: List[str] = []
    frameworks: List[str] = []
    libraries: List[str] = []
    databases: List[str] = []
    cloud: List[str] = []
    tools: List[str] = []

class ExperienceEntrySchema(BaseModel):
    id: Optional[str] = None
    company: str = ""
    role: str = ""
    duration: str = ""
    description: str = ""

class ProjectEntrySchema(BaseModel):
    id: Optional[str] = None
    title: str = ""
    technologies: List[str] = []
    description: str = ""

class ParsedResumeSchema(BaseModel):
    personalInfo: PersonalInformationSchema = Field(default_factory=PersonalInformationSchema)
    education: List[EducationEntrySchema] = []
    skills: CategorizedSkillsSchema = Field(default_factory=CategorizedSkillsSchema)
    experience: List[ExperienceEntrySchema] = []
    projects: List[ProjectEntrySchema] = []
    certifications: List[str] = []
    achievements: List[str] = []
    languages: List[str] = []

class CategoryScoreDetail(BaseModel):
    score: float
    max: float
    details: str

class ATSCategoryBreakdownSchema(BaseModel):
    contactInfo: CategoryScoreDetail
    education: CategoryScoreDetail
    projects: CategoryScoreDetail
    skills: CategoryScoreDetail
    experience: CategoryScoreDetail
    keywords: CategoryScoreDetail
    formatting: CategoryScoreDetail

class ATSScoreResponseSchema(BaseModel):
    overallScore: float
    breakdown: ATSCategoryBreakdownSchema
    grade: str
    statusText: str

class HealthReportItemSchema(BaseModel):
    category: str
    title: str
    description: str
    type: str

class HealthReportSchema(BaseModel):
    strengths: List[str] = []
    weaknesses: List[str] = []
    missingSections: List[str] = []
    duplicateInfo: List[str] = []
    grammarAlerts: List[str] = []
    keywordDensityRating: str = "Moderate"
    formattingQuality: str = "Good"
    overallReadabilityScore: float = 80.0
    items: List[HealthReportItemSchema] = []

class SuggestionSchema(BaseModel):
    id: str
    category: str
    problem: str
    reason: str
    current: str
    suggested: str
    impactScore: Optional[float] = 10.0

class JobMatchRequestSchema(BaseModel):
    jobId: Optional[str] = None
    jobDescription: Optional[str] = None
    resumeId: Optional[str] = None

class JobMatchResultSchema(BaseModel):
    jobId: Optional[str] = None
    jobTitle: str
    companyName: str
    matchPercentage: float
    satisfiedSkills: List[str] = []
    missingSkills: List[str] = []
    educationMatch: bool = True
    experienceMatch: bool = True
    matchingKeywords: List[str] = []
    recommendationReason: str = ""

class SkillGapAnalysisSchema(BaseModel):
    matchPercentage: float
    alreadyKnown: List[str] = []
    needToLearn: List[str] = []
    prioritySkills: List[str] = []
    optionalSkills: List[str] = []
    gapProgress: float = 0.0

class JobRecommendationSchema(BaseModel):
    id: str
    jobTitle: str
    companyName: str
    location: str
    salary: str
    matchPercentage: float
    tier: str
    reason: str
    matchingSkills: List[str] = []
    missingSkills: List[str] = []
    expectedAtsScore: float

class ResumeVersionSchema(BaseModel):
    id: str
    userId: str
    filename: str
    originalName: str
    fileSize: int
    fileType: str
    atsScore: float
    isActive: bool
    uploadedAt: str

class ResumeAnalyticsSchema(BaseModel):
    skillDistribution: List[Dict[str, Any]] = []
    atsTrend: List[Dict[str, Any]] = []
    applicationsCount: int = 0
    resumeImprovementRate: float = 0.0
    jobMatchesCount: int = 0
    monthlyUploads: List[Dict[str, Any]] = []

class ActiveResumeResponseSchema(BaseModel):
    id: str
    userId: str
    filename: str
    originalName: str
    fileUrl: str
    fileSize: int
    fileType: str
    parsedData: ParsedResumeSchema
    atsScore: float
    atsBreakdown: ATSCategoryBreakdownSchema
    healthReport: HealthReportSchema
    suggestions: List[SuggestionSchema] = []
    isActive: bool
    uploadedAt: str
    versions: List[ResumeVersionSchema] = []

File: packages/types/src/resume.ts
export interface PersonalInformation {
  name: string;
  email: string;
  phone: string;
  linkedin: string;
  github: string;
  portfolio: string;
  location?: string;
  headline?: string;
  bio?: string;
}

export interface EducationEntry {
  id?: string;
  degree: string;
  college: string;
  cgpa: string;
  graduationYear: string;
  fieldOfStudy?: string;
}

export interface CategorizedSkills {
  programmingLanguages: string[];
  frameworks: string[];
  libraries: string[];
  databases: string[];
  cloud: string[];
  tools: string[];
}

export interface ExperienceEntry {
  id?: string;
  company: string;
  role: string;
  duration: string;
  description: string;
  technologies?: string[];
}

export interface ProjectEntry {
  id?: string;
  title: string;
  technologies: string[];
  description: string;
  liveUrl?: string;
  repoUrl?: string;
}

export interface ParsedResume {
  personalInfo: PersonalInformation;
  education: EducationEntry[];
  skills: CategorizedSkills;
  experience: ExperienceEntry[];
  projects: ProjectEntry[];
  certifications: string[];
  achievements: string[];
  languages: string[];
}

export interface ATSCategoryBreakdown {
  contactInfo: { score: number; max: 10; details: string };
  education: { score: number; max: 15; details: string };
  projects: { score: number; max: 20; details: string };
  skills: { score: number; max: 25; details: string };
  experience: { score: number; max: 15; details: string };
  keywords: { score: number; max: 10; details: string };
  formatting: { score: number; max: 5; details: string };
}

export interface ATSScoreResult {
  overallScore: number;
  breakdown: ATSCategoryBreakdown;
  grade: 'Red' | 'Yellow' | 'Green';
  statusText: string;
}

export interface HealthReportItem {
  category: string;
  title: string;
  description: string;
  type: 'strength' | 'weakness' | 'missing' | 'warning' | 'info';
}

export interface HealthReport {
  strengths: string[];
  weaknesses: string[];
  missingSections: string[];
  duplicateInfo: string[];
  grammarAlerts: string[];
  keywordDensityRating: 'Low' | 'Moderate' | 'Optimal' | 'Overstuffed';
  formattingQuality: 'Basic' | 'Good' | 'Excellent';
  overallReadabilityScore: number; // 0-100
  items: HealthReportItem[];
}

export interface AiSuggestion {
  id: string;
  category: string;
  problem: string;
  reason: string;
  current: string;
  suggested: string;
  impactScore?: number;
}

export interface SkillGapCategory {
  categoryName: string;
  skills: string[];
}

export interface SkillGapAnalysis {
  matchPercentage: number;
  alreadyKnown: string[];
  needToLearn: string[];
  prioritySkills: string[];
  optionalSkills: string[];
  gapProgress: number; // 0-100%
}

export interface JobMatchResult {
  jobId?: string;
  jobTitle: string;
  companyName: string;
  matchPercentage: number;
  satisfiedSkills: string[];
  missingSkills: string[];
  educationMatch: boolean;
  experienceMatch: boolean;
  matchingKeywords: string[];
  recommendationReason: string;
}

export interface JobRecommendation {
  id: string;
  jobTitle: string;
  companyName: string;
  location: string;
  salary: string;
  matchPercentage: number;
  tier: 'Top Match' | 'Good Match' | 'Stretch Match';
  reason: string;
  matchingSkills: string[];
  missingSkills: string[];
  expectedAtsScore: number;
}

export interface ResumeVersion {
  id: string;
  userId: string;
  filename: string;
  originalName: string;
  fileSize: number;
  fileType: string;
  atsScore: number;
  isActive: boolean;
  versionNumber?: number;
  uploadedAt: string;
}

export interface ResumeAnalytics {
  skillDistribution: { category: string; count: number }[];
  atsTrend: { date: string; score: number; version: string }[];
  applicationsCount: number;
  resumeImprovementRate: number;
  jobMatchesCount: number;
  monthlyUploads: { month: string; uploads: number }[];
}

export interface ActiveResumeResponse {
  id: string;
  userId: string;
  filename: string;
  originalName: string;
  fileUrl: string;
  fileSize: number;
  fileType: string;
  parsedData: ParsedResume;
  atsScore: number;
  atsBreakdown: ATSCategoryBreakdown;
  healthReport: HealthReport;
  suggestions: AiSuggestion[];
  isActive: boolean;
  versionNumber?: number;
  uploadedAt: string;
  versions: ResumeVersion[];
}
